File: Java-Core/src/main/java/prompto/grammar/ArgumentAssignmentList.java
package prompto.grammar;

import java.util.LinkedList;

import prompto.declaration.IMethodDeclaration;
import prompto.error.SyntaxError;
import prompto.runtime.Context;
import prompto.utils.CodeWriter;


public class ArgumentAssignmentList extends LinkedList<ArgumentAssignment> {

	private static final long serialVersionUID = 1L;

	public ArgumentAssignmentList() {
		
	}
	
	public ArgumentAssignmentList(ArgumentAssignment assignment) {
		this.add(assignment);
	}
	
	public ArgumentAssignmentList(ArgumentAssignmentList assignments) {
		super(assignments);
	}

	public ArgumentAssignmentList(ArgumentAssignmentList assignments, ArgumentAssignment assignment) {
		super(assignments);
		this.add(assignment);
	}

	public ArgumentAssignment find(Identifier name) {
		for(ArgumentAssignment assignment : this) {
			if(name.equals(assignment.getName()))
				return assignment;
		}
		return null;
	}

	public ArgumentAssignmentList makeAssignments(Context context, IMethodDeclaration declaration) throws SyntaxError {
		ArgumentAssignmentList assignments = new ArgumentAssignmentList();
		for(ArgumentAssignment assignment : this)
			assignments.add(assignment.makeAssignment(context, declaration));
		return assignments;
	}
	
	public void toDialect(CodeWriter writer) {
		switch(writer.getDialect()) {
		case E:
			toEDialect(writer);
			break;
		case O:
			toODialect(writer);
			break;
		case S:
			toPDialect(writer);
			break;
		}
	}
	
	private void toEDialect(CodeWriter writer) {
		int idx = 0;
		// anonymous argument before 'with'
		if(this.size()>0 && this.get(0).getArgument()==null) {
			writer.append(' ');
			this.get(idx++).toDialect(writer);
		}
		if(idx<this.size()) {
			writer.append(" with ");
			this.get(idx++).toDialect(writer);
			writer.append(", ");
			while(idx<this.size()-1) {
				this.get(idx++).toDialect(writer);
				writer.append(", ");
			}
			writer.trimLast(2);
			if(idx<this.size()) {
				writer.append(" and ");				
				this.get(idx++).toDialect(writer);
			}
		}
	}
	
	private void toODialect(CodeWriter writer) {
		writer.append("(");
		for(ArgumentAssignment as : this) {
			as.toDialect(writer);
			writer.append(", ");
		}
		if(this.size()>0)
			writer.trimLast(2);
		writer.append(")");
	}

	private void toPDialect(CodeWriter writer) {
		toODialect(writer);
	}


}

File: Java-CodeServer/src/main/resources/js/prompto/grammar/ArgumentAssignmentList.js
var ObjectList = require("../utils/ObjectList").ObjectList;
var Dialect = require("../parser/Dialect").Dialect;

function ArgumentAssignmentList(items, item) {
	items = items || [];
	ObjectList.call(this, items);
	item = item || null;
	if(item!==null) {
		this.add(item);
	}
	return this;
}

ArgumentAssignmentList.prototype = Object.create(ObjectList.prototype);
ArgumentAssignmentList.prototype.constructor = ArgumentAssignmentList;
	

ArgumentAssignmentList.prototype.findIndex = function(name) {
	for(var i=0;i<this.length;i++) {
		if(name==this[i].name) {
			return i;
		}
	}
	return -1;
};

ArgumentAssignmentList.prototype.find = function(name) {
	for(var i=0;i<this.length;i++) {
		if(name==this[i].name) {
			return this[i];
		}
	}
	return None;
};

ArgumentAssignmentList.prototype.makeAssignments = function(context, declaration) {
	var assignments = new ArgumentAssignmentList();
	for(var i=0;i<this.length;i++) {
		assignments.push(this[i].makeAssignment(context, declaration));
	}
	return assignments;
};



ArgumentAssignmentList.prototype.toDialect = function(writer) {
    writer.toDialect(this);
};

ArgumentAssignmentList.prototype.toEDialect = function(writer) {
    var idx = 0;
    // anonymous argument before 'with'
    if(this.length>0 && this[0].argument==null) {
        writer.append(' ');
        this[idx++].toDialect(writer);
    }
    if(idx<this.length) {
        writer.append(" with ");
        this[idx++].toDialect(writer);
        writer.append(", ");
        while(idx<this.length-1) {
            this[idx++].toDialect(writer);
            writer.append(", ");
        }
        writer.trimLast(2);
        if(idx<this.length) {
            writer.append(" and ");
            this[idx++].toDialect(writer);
        }
    }
};

ArgumentAssignmentList.prototype.toODialect = function(writer) {
    writer.append("(");
    for(var i=0; i<this.length; i++) {
        this[i].toDialect(writer);
        writer.append(", ");
    }
    if(this.length>0)
        writer.trimLast(2);
    writer.append(")");
};

ArgumentAssignmentList.prototype.toSDialect = function(writer) {
    this.toODialect(writer);
};


exports.ArgumentAssignmentList = ArgumentAssignmentList;

